include/isc_nav/utility/rrt.hpp

```diff
@@ -0,0 +1,267 @@
+                {
+                    Xnew = findNew(Xnearest, random, Xnearest_dist);
+                }
+
+                if (isInObstacle(Xnew))
+                {
+                    continue;
+                }
+                
+                // Make sure no obstacle between Xnearest and Xnew
+                if (!isCollisionFree(Xnearest, Xnew))
+                {
+                    continue;
+                }
+
+                // Get the index of to and from
+                int attach_to = getIndex(V, Xnew);
+                int attach_from = getIndex(V, Xnearest);
+
+                // Add edge
+                if (attach_to == -1)
+                {
+                    int add_new = V.size();
+                    V.push_back(Xnew);
+                    E.push_back(0);
+
+                    for (int i = 0; i < attach_to; i++)
+                    {
+                        G.at(i).push_back(0);
+                    }
+
+                    G.at(attach_from).at(add_new)++;
+                    G.push_back(E);
+                }
+                else 
+                {
+                    V.push_back(Xnew);
+                    E.push_back(0);
+
+                    for (int i = 0; i < attach_to; i++)
+                    {
+                        G.at(i).push_back(0);
+                    }
+
+                    G.at(attach_from).at(attach_to)++;
+                    G.push_back(E);
+                }
+
+                // If vertex at goal return graph
+                if ((Xnew.x - goal_.x < 0.1) && (Xnew.y - goal_.y < 0.1))
+                {
+                    trace_back_path(Xnew, path);
+                }
+            }
+
+            return path;
+        }
+
+
+        Point2D randomPosition()
+        {
+            srand(time(0));
+
+            while (true) 
+            {
+                int32_t x = fRand(0, map_.get_size_x());
+                int32_t y = fRand(0, map_.get_size_y());
+
+                if (map_.valid_cell(x, y)) 
+                {
+                    Point2D random_node(x, y);
+                    return random_node;
+                }
+            }
+        }
+
+
+        Point2D findNearest(std::vector<Point2D>& V, Point2D random_node, double& Xnearest_dist)
+        {
+            double dist;
+            Point2D Xnearest;
+
+            if (!V.empty()) 
+            {
+                // Assign values to be compared
+                Xnearest_dist = getDistance(V.at(0), random_node);
+                Xnearest = V.at(0);
+            }
+
+            for (auto vertex : V)
+            {
+                dist = getDistance(vertex, random_node);
+                if (dist < Xnearest_dist)
+                {
+                    Xnearest_dist = dist;
+                    Xnearest = vertex;
+                }
+            }
+
+            return Xnearest;
+        }
+
+        // p1 is Xnearest, p2 is random node
+        Point2D findNew(Point2D& p1, Point2D& p2, double& Xnearest_dist)
```

Comment:
This applies across C++, when passing by refrence, always const unless the variable is actually being mutated. This is done so you don't accidentally update an arguement in the function, which goes a long way to help with debugging.

in this example `const double& Xnearest_dist`. Although here the const ref actually isnt doing anything as its just as expensive as a pass by value as both the refrenence and the value are 64 bits (it is worth it for the points though).